File: game/banker.py
from .config import GameConfig


class Banker:
    def __init__(self, config: GameConfig | None = None) -> None:
        self._config = config or GameConfig()

    def make_offer(
        self,
        unopened_amounts: list[int],
        completed_offer_rounds: int,
    ) -> int:
        if not unopened_amounts:
            return 0

        discount_index = min(
            completed_offer_rounds,
            len(self._config.offer_discounts) - 1,
        )
        average = sum(unopened_amounts) / len(unopened_amounts)
        return max(0, round(average * self._config.offer_discounts[discount_index]))

File: flutter_android/lib/game/banker.dart
import 'game_config.dart';
import 'money_formatter.dart';

class Banker {
  const Banker();

  int makeOffer({
    required Iterable<int> unopenedAmounts,
    required int completedOfferRounds,
  }) {
    final amounts = unopenedAmounts.toList(growable: false);
    if (amounts.isEmpty) {
      return 0;
    }

    final total = amounts.fold<int>(0, (sum, value) => sum + value);
    final average = total / amounts.length;
    final discountIndex = completedOfferRounds.clamp(
      0,
      GameConfig.offerDiscounts.length - 1,
    ) as int;
    final discount = GameConfig.offerDiscounts[discountIndex];

    return clampOffer(average * discount);
  }
}
